lib/ABI/FunctionType.cpp

```diff
@@ -52,74 +55,95 @@ bool verify(const SortedVector<RegisterType> &UsedRegisters,
+template<model::ABI::Values ABI>
+static constexpr bool verifyReturnValueLocationHelper() {
+  using AT = abi::Trait<ABI>;
+  constexpr auto &LocationRegister = AT::ReturnValueLocationRegister;
 
-static std::optional<model::QualifiedType>
-buildDoubleType(model::Register::Values UpperRegister,
-                model::Register::Values LowerRegister,
-                model::PrimitiveTypeKind::Values CustomKind,
-                model::Binary &TheBinary) {
-  model::PrimitiveTypeKind::Values UpperKind = selectTypeKind(UpperRegister);
-  model::PrimitiveTypeKind::Values LowerKind = selectTypeKind(LowerRegister);
-  if (UpperKind != LowerKind)
-    return std::nullopt;
+  // Skip ABIs that do not allow returning big values.
+  // They do not benefit from this check.
+  if constexpr (LocationRegister != model::Register::Invalid) {
+    if constexpr (isVectorRegisterHelper(LocationRegister)) {
+      // Vector register used as the return value locations are not supported.
+      return false;
+    } else if constexpr (constexprIsContained(AT::CalleeSavedRegisters,
+                                              LocationRegister)) {
+      // Using callee saved register as a return value location doesn't make
+      // much sense: filter those out.
+        return false;
+    } else {
+      // The return value location register can optionally also be the first GPRs,
+      // but only the first one.
+      constexpr auto &GPRs = AT::GeneralPurposeArgumentRegisters;
+      constexpr auto Iterator = constexprFind(GPRs, LocationRegister);
+      if constexpr (Iterator != GPRs.end() && Iterator != GPRs.begin())
+        return false;
+    }
+  }
 
-  size_t UpperSize = model::Register::getSize(UpperRegister);
-  size_t LowerSize = model::Register::getSize(LowerRegister);
-  return model::QualifiedType(TheBinary.getPrimitiveType(CustomKind,
-                                                         UpperSize + LowerSize),
-                              {});
+  return true;
 }
 
-static model::QualifiedType getTypeOrDefault(const model::QualifiedType &Type,
-                                             model::Register::Values Register,
-                                             model::Binary &Binary) {
-  if (Type.UnqualifiedType.get() != nullptr)
-    return Type;
-  else
-    return buildType(Register, Binary);
-}
+/// \brief Replace all the references to `OldKey` with the references to
+///        the newly added `NewType`. Also erases the original type.
+///
+/// \param OldKey The type references of which should be replaced.
+/// \param NewType The new type to replace the references to. It is added to
+///        the model implicitly, no need to do so before calling this function.
+/// \param Model The tuple tree where replacement should take place in.
+///
+/// \return The new path to the added type.
+static model::TypePath
+fullTypeReplacement(const model::Type::Key &OldKey,
+                    UpcastablePointer<model::Type> &&NewType,
```

Comment:
Maybe it's better to let the caller insert the type.

Reply:
Yeah, I really hesitated about this. But in the end I came to the conclusion that leaving the model without the type that's referred to all over the place is kind of dangerous. So I'd rather put inserting and erasure as close together as possible. Think of this function as a safe and self-contained black box that makes sure that one of the types is safely replaced by another one (including all the references) - and nothing else.

Reply:
Uhm I mean, the caller does the following:
1. Create the new type
2. Insert it in the model
3. `fullTypeReplacement`
4. Drop the old type

Reply:
Thanks to `Binary.newType<>()`, this is less necessary and such this thread is no longer relevant.